packages/ra-core/src/dataProvider/withLifecycleCallbacks.ts

```diff
@@ -574,20 +466,48 @@ export type ResourceCallbacks<T extends RaRecord = any> = {
      * Note: This callback doesn't modify the record itself, but the data argument
      * (which may be a diff, especially when called with updateMany).
      */
-    beforeSave?: (
+    beforeSave?: ((
         data: Partial<T>,
-        dataProvider: DataProvider
-    ) => Promise<Partial<T>>;
+        utils: ResourceCallbackUtils
+    ) => Promise<Partial<T>>)[];
     /**
      * Update a record after it has been read from the dataProvider
      *
      * Used in getOne, getList, getMany, and getManyReference
      */
-    afterRead?: (record: T, dataProvider: DataProvider) => Promise<T>;
+    afterRead?: ((record: T, utils: ResourceCallbackUtils) => Promise<T>)[];
     /**
      * Use the record after it is returned by the dataProvider.
      *
      * Used in create, update, and updateMany
      */
-    afterSave?: (record: T, dataProvider: DataProvider) => Promise<T>;
+    afterSave?: ((record: T, utils: ResourceCallbackUtils) => Promise<T>)[];
 };
+
+/**
+ * The list of all possible data provider hooks
+ * @see https://marmelab.com/react-admin/DataProviders.html#data-provider-hooks
+ */
+export const dataProviderHooks = [
```

Comment:
Why do you export it?

Reply:
I use it in my own code to apply a callback for every possible hook (for login purpose essentially).
I understand this could be in my own code though.